ProjFS.Mac/PrjFSKext/PrjFSKext/AssociativeArray.hpp

```diff
@@ -0,0 +1,203 @@
+        uint32_t expectedMemorySize = this->MemorySizeToGrow();
+        if (memorySize != expectedMemorySize)
+        {
+            return ResizeResult{ false };
+        }
+        
+        return this->ReserveCapacity(expectedCapacity, memory, memorySize);
+    }
+
+    ResizeResult ReserveCapacity(uint32_t newCapacity, void* memory, uint32_t memorySize)
+    {
+        assert(MemorySizeForCapacity(newCapacity) == memorySize);
+        assert(newCapacity >= this->numberUsed);
+        
+        uint8_t* memoryPosition = static_cast<uint8_t*>(memory);
+        for (size_t i = 0; i < this->numberUsed; ++i, memoryPosition += sizeof(VALUE))
+        {
+            VALUE* moved = new(memoryPosition) VALUE(KextCpp::move(this->data[i]));
+            (void)moved;
+            this->data[i].~VALUE();
+        }
+        
+        ResizeResult result = { true, MemorySizeForCapacity(this->capacity), this->data };
+        this->data = static_cast<VALUE*>(memory);
+        this->capacity = memorySize / sizeof(VALUE);
+        return result;
+    }
+    
+    VALUE* Find(KEY key)
+    {
+        for (size_t i = 0; i < this->numberUsed; ++i)
+        {
+            if (EQUAL_FN(key, this->data[i]))
+            {
+                return &this->data[i];
+            }
+        }
+        
+        return nullptr;
+    }
+    
+    // Inserts a new item, if there is space. No checks for duplicates are done.
+    VALUE* TryInsert(VALUE&& newItem)
+    {
+        size_t nextIndex = this->numberUsed;
+        if (nextIndex < this->capacity)
+        {
+            ++this->numberUsed;
+            VALUE* inserted = new(&this->data[nextIndex]) VALUE(newItem);
+            assert(inserted == &this->data[nextIndex]);
+            return inserted;
+        }
+        else
+        {
+            return nullptr;
+        }
+    }
+    
+    VALUE* FindOrTryInsert(KEY key, VALUE&& newItem)
+    {
+        VALUE* value = this->Find(key);
+        if (nullptr == value)
+        {
+            value = this->TryInsert(KextCpp::move(newItem));
+        }
+        return value;
+    }
+    
+    void Remove(VALUE* item)
+    {
+        assert(item != nullptr);
+        assert(item >= this->data && item < this->data + this->capacity);
+        assert(this->numberUsed > 0);
+        
+        // move the last item into the liberated slot if not removing the last item
+        VALUE* lastItem = this->data + (this->numberUsed - 1);
+        if (lastItem != item)
+        {
+            *item = KextCpp::move(*lastItem);
+        }
+        lastItem->~VALUE();
+
+        --this->numberUsed;
+    }
+    
+    // Find item matching key; if not found, insert newItem, growing the array if necessary
+    // (during which the lock may be dropped and then re-acquired (exclusively))
+    // Returns the found or inserted value, or nullptr if resizing failed.
+    VALUE* FindOrInsertLockedAllowResize(KEY key, VALUE&& newItem, RWLock& lock)
```

Comment:
This is a direct consequence of avoiding memory allocations while holding a lock. If this were not the kernel, we'd do something like:

 * Lock `s_virtualizationRootsLock` with exclusive access.
 * Call a `FindOrInsert()` function on the data structure.
 * Item not found, so insert.
 * No space left, so resize.
 * **Allocate some memory** to hold the resized array.
 * Move items across.
 * Perform the insertion.
 * Return from `FindOrInsert()`
 * Unlock

Note how this allocates memory while a lock is held. So in a kext, I'd implement it as this instead:

 * Lock `s_virtualizationRootsLock` with exclusive access.
 * Call a `FindOrInsert()` function on the data structure.
 * Item not found, so insert.
 * No space left, so attempt to resize.
 * **Try allocating some memory with non-blocking allocation** to hold the resized array<br>(*Assume for illustrative purposes the allocation fails; this doesn't always happen of course, but I'm showing the worst case.*)
 * Record how much memory we need for resizing.
 * Drop the lock.
 * **Allocate memory required for resizing using *blocking* alloc.**
 * Re-acquire the lock.
 * Search for the item again. (NB: Another thread may have inserted it since.)
 * Item not found, so try to insert. (NB: Another thread may have resized the data structure or removed items since we last checked for space.)
 * No space left, try to resize using the memory we allocated while not holding the lock.<br>(*This could fail if another thread already resized and then filled all the new space, so our allocated memory is no longer large enough; so you actually have to attempt this in a loop.*)
 * We have the right amount of memory, go ahead with resizing.
 * Move items across.
 * Perform the insertion.
 * Return from `FindOrInsert()`
 * Unlock

This gets extremely complicated if we only allow dropping the lock in the calling code; the calling code will then have to deal with the memory allocation while dropping the lock, explicitly perform the resize, then try inserting again, etc.

Sometimes you *have* to make the calling code do the leg work - this is the case if the insertion and another operation must be atomic, so the lock can't simply be dropped at will. Where that's not a problem though, passing the lock that needs dropping during allocations into the data structure mutation function is by far the most straightforward.

Reply:
> Where that's not a problem though, passing the lock that needs dropping during allocations into the data structure mutation function is by far the most straightforward.
> This should hopefully also answer your other questions about seemingly odd API design.

All that makes sense, and thanks again for the details.  While reviewing the changes with @jeschu1 this morning we were wondering about the exact scenario you call out here:

> Sometimes you have to make the calling code do the leg work - this is the case if the insertion and another operation must be atomic, so the lock can't simply be dropped at will. 

But it sounds like that's a complexity you (as a the author of the calling code) must be aware of and make sure that if you're using this function it is safe to have the lock dropped and reacquired.

Reply:
> But it sounds like that's a complexity you (as a the author of the calling code) must be aware of and make sure that if you're using this function it is safe to have the lock dropped and reacquired.

Yes, I was on the fence on whether to add a rather clumsy `MayDropLock` suffix to the function name to be super explicit about it. Any better suggestions?

Reply:
@pmj, after revisiting this code (and reading the update on your other comment), I'm now wondering if we need the complexity of dropping the lock when increasing the size of `s_usedMountPoints`.

Could we instead add a new `s_usedMountPointsLock` and use that when modifying `s_usedMountPoints`?

Because `s_usedMountPoints` is not accessed during any of the kext callbacks, I'm wondering if using a second lock would allow us to avoid deadlocks without the complexity of dropping\re-acquiring the lock.  

We would not need to worry about the allocator paging out memory because if it does (and our kext is called) none of those code paths would acquire `s_usedMountPointsLock`.

Reply:
Yes, this is a good point, we can completely isolate that aspect, as we're not attempting to make the auth cache disabling atomic with regard to provider registration. In any case, we could always increment *before* registering the provider, and decrement in case it fails. And yes, it *should* be safe to alloc memory while holding that lock, although this sort of contradicts the philosophy you floated yesterday, that you were thinking we shouldn't have generic data structures as it would make memory allocations too easy for new devs. In the same vein, I would say that allocating memory while holding this newly introduced lock would possibly make it too easy for new code to be introduced where the lock is acquired in a context where those allocations *could* cause a deadlock.

So I think adding and using the new lock is definitely a good idea simply because we *can* do it, but if we're going with dynamic allocations, I'd still vote against allocating while the lock is held.